Overhaul the entire library

kleinesfilmroellchen opened the pull request:
- Port to esp-hal 1.0 (version is pinned, as RMT is unstable and might change in every new release)
- Allow flexibly specifying LED timing modes and color order
- Implement common timings and color orders
- Improve dependency specification
- Change buffer size macro to remove the requirement to pre-allocate a stack buffer (when this can be done in the adapter directly)
- Fix both examples
- Add more documentation
- Remove documented-features as all the features are for development and examples only
- Improve docs-rs situation
- Make error type non-exhaustive so adding more errors won’t be a semver-breaking change
- Use edition 2024
- Align a lot of code more closely with common Rust practices
- Add async interface: when Rmt/Channel are async, we implement `SmartLedWriteAsync`.
- Example for async interface

I hope these changes are acceptable. They make the library more flexible and less error-prone to use, while requiring only minor changes for most users (different construction of the driver). The fundamental driver implementation hasn’t changed all that much.

For now I will publish this crate on crates.io as well, since there hasn’t been activity here since the esp-hal 1.0 release.

kleinesfilmroellchen committed:
Overhaul the entire library

- Port to esp-hal 1.0 (version is pinned, as RMT is unstable and might change in every new release)
- Allow flexibly specifying LED timing modes and color order
- Implement common timings and color orders
- Improve dependency specification
- Change buffer size macro to remove the requirement to pre-allocate a stack buffer (when this can be done in the adapter directly)
- Fix both examples
- Add more documentation
- Remove documented-features as all the features are for development and examples only
- Improve docs-rs situation
- Make error type non-exhaustive so adding more errors won’t be a semver-breaking change

kleinesfilmroellchen committed:
fix broken documentation

kleinesfilmroellchen committed:
prepare for crates.io

kleinesfilmroellchen committed:
Slight improvement to error handling during transmission

note about broken transmit error case -> known upstream problem, will be fixed soon

kleinesfilmroellchen committed:
add required metadata for crates.io

kleinesfilmroellchen committed:
fix build targets for docs.rs

kleinesfilmroellchen committed:
bump version

kleinesfilmroellchen committed:
add license and description

kleinesfilmroellchen committed:
Implement async functionality

This is actually a lot simpler than non-async and doesn’t require the Option<Channel> workaround.

Add example for async with embassy, which matches the existing example but also has a background task.

kleinesfilmroellchen committed:
bump version again

kleinesfilmroellchen committed:
Adjust WS2811 timings

What we used previously was only the low-speed mode

kleinesfilmroellchen committed:
bump version

TuxCoder commented:
Hello, was searching for something like your PR and found it on crates.io
So I testet it. My project plans include an esp32-c6 but that is in shipping.
But have had a ESP32 around (with the bit annoying xtensa rust toolchain).

So Problem there is "esp" rust latest release for xtensa is 1.90  https://github.com/esp-rs/rust

I don't think you need 1.91, can you change that to 1.90.

Otherwise look pretty good, good work!

Otherwise I have some strange RMT bugs,
debugging, ....
signal is strange after 2.333 LEDS ( 56bits ), and is longer as it should be ~

edit: got it working, will do a patch

kleinesfilmroellchen commented:
It is known that RMT dividers >10 can cause issues, see https://github.com/esp-rs/esp-hal/issues/3930#issuecomment-3486677713 . The MSRV is now fixed in 0.25.1.

kleinesfilmroellchen committed:
Decrement required MSRV

allows xtensa boards again

TuxCoder committed:
fix RMT issues

this fixes issues in the RMT-DMA transfere

memsize uses two channels to do kind of double buffering.

kleinesfilmroellchen committed:
Make memsize specification optional

also add proper error handling, which is more important with extra memsize

kleinesfilmroellchen committed:
Update docs references to point at smartled2

kleinesfilmroellchen committed:
Expand documentation even more

kleinesfilmroellchen committed:
Add migration guide

kleinesfilmroellchen commented:
Closing, as it doesn’t seem like this will lead to anything and keeping the commits separated is a lot of effort. @TuxCoder your changes have been incorporated in a modified way into 0.26, since unconditional use of two memory regions prevents other RMT channels from working.

TuxCoder commented:
Thanks for the patches!

I poste here as it looks like a led driver specific issue and not a RMT specific,
and you have not enabled issues on your repo (yet, only do if you want to maintain a new project :) ).

For now I post my findings here.

I did some testing and this is my current interpretation.

For the ESP32 I needed to increase the `memsize` as otherwise, it looked like the cpu was not fast enough in feeding data into the RMT buffer (probably IRQ handling / ...) and started to repeat patterns instant of sending the next one.
for the ESP32 this is 64Pulses https://docs.espressif.com/projects/rust/esp-metadata-generated/0.2.0/esp32/src/esp_metadata_generated/_generated_esp32.rs.html#183

After increasing `memsize` I can run 300leds without a problem.

Looks like a Hardware limitation,
so would probably make sense to ether increase the `memsize` for all devices or make it configurable.
With memsize(2) only every second channel can be allocated, what half the usable units.
So not sure what would be a good solution, but it works for me ?!.

I think fastled also uses double buffers:
https://github.com/FastLED/FastLED/blob/master/src/platforms/esp/32/drivers/rmt/README_RMT_WIFI_ROBUSTNESS.md?plain=1#L37


I also tested a bit with multiple LED strips.
I did this with a `embassy-futures::join::join`
If I use the current API it does not work and one output does not get data correctly.

As my understanding from rust future / async is, that a futures only get executed if awaited,
one output buffer is prepared and started to send,
then the other futures get started to get executed, what also first prepares the buffer and starts then to send.
This blocks the cpu for a good amount of time, what starves the first buffer.

I did some test implementation to seperate prepare buffer and send buffer.
From my understanding this is not posible with the `SmartLedsWriteAsync` trait, so did a bit hack around.

Result is here:

Data is a counter for easier debugging, led count is 300 * 2

Beginning sequence:
<img width="3101" height="572" alt="image" src="https://github.com/user-attachments/assets/e28835f7-9376-49fb-88d5-7fdfbcea63ee" />

Ending sequence:
<img width="2070" height="423" alt="image" src="https://github.com/user-attachments/assets/066484e5-2086-4682-b17f-888043cfe246" />


Will publish code soon, but busy with child and need a bit cleanup, started the project 1 day ago.

TuxCoder commented:
I found a solution to run multiple long led strips.

If I slipt the `async fn write()` into a creation part and a send part, we can create multiple futures with prepared data,
and send them then all in sync.

Edit, did a rebase:
https://github.com/TuxCoder/esp-hal-smartled/commit/b24e45bfbd3b3be558818460d049c6730b2f4620

I could run with that 4 channels in sync with each 300LEDS.

testcode:
```rust
    loop {
        let now = Instant::now();
        info!("loop {}", now.as_millis());

        // creates an iterator for NUM_LEDS leds
        let animation = TestCounter::<NUM_LEDS>::new();

        let led0_f = led0.write(animation.clone());
        let led1_f = led1.write(animation.clone());
        let led2_f = led2.write(animation.clone());
        let led3_f = led3.write(animation.clone());
        let (led0_r, led1_r, led2_r, led3_r) = join::join4(led0_f, led1_f, led2_f, led3_f).await;
        led0_r.unwrap();
        led1_r.unwrap();
        led2_r.unwrap();
        led3_r.unwrap();
    }
```

start sequence:
<img width="1003" height="628" alt="image" src="https://github.com/user-attachments/assets/50eca93f-57e1-40ea-a2be-552b1d044b36" />

end sequence:
<img width="1436" height="622" alt="image" src="https://github.com/user-attachments/assets/15526c2e-be72-411a-bf80-735efabb9561" />

kleinesfilmroellchen commented:
> Thanks for the patches!
> 
> I poste here as it looks like a led driver specific issue and not a RMT specific, and you have not enabled issues on your repo (yet, only do if you want to maintain a new project :) ).

Fixed! GitHub appears to have done this automatically. Feel free to write up any issues you found on my repo.

> Looks like a Hardware limitation, so would probably make sense to ether increase the `memsize` for all devices or make it configurable. With memsize(2) only every second channel can be allocated, what half the usable units. So not sure what would be a good solution, but it works for me ?!.

Original ESP32 is indeed quite a bit slower and less capable (Xtensa LX6 vs. LX7 on S2/S3). The channels are also smaller (should be 48*2 channels at minimum for newer devices, e.g. C6).

I think the memsize is a fine solution for these cases. At least on models like the C6 you can even increase it up to 4, which then also occupies the two receive channel buffers. But as we observed it’s a trade-off and not necessary on faster chips and fewer LEDs, which is why I made it an optional configuration. I assume that this worked for you, as released in 0.26?

> I think fastled also uses double buffers: https://github.com/FastLED/FastLED/blob/master/src/platforms/esp/32/drivers/rmt/README_RMT_WIFI_ROBUSTNESS.md?plain=1#L37

Double-buffering requires upstream support. As you know @whisperwind is working on a solution, and once one becomes available I’ll look into incorporating it (or you can do so and send a PR).

> I also tested a bit with multiple LED strips. I did this with a `embassy-futures::join::join` If I use the current API it does not work and one output does not get data correctly.
> 
> As my understanding from rust future / async is, that a futures only get executed if awaited, one output buffer is prepared and started to send, then the other futures get started to get executed, what also first prepares the buffer and starts then to send. This blocks the cpu for a good amount of time, what starves the first buffer.

Yes, that’s expected, I’ve seen this with a different project too. If you used separate tasks, this may work better. This is also how the async transmit function works: it only starts the transmission once awaited. It could be implemented differently (without the `async` sugar, which is something I hacked together for my use-case too, and which is probably exactly like the thing you did), but there appears to be some waiting time required (~20-50ms) when dispatching two different channels, possibly enforced by hardware. The only real fix is to do simultaneous TX, which some chips have in hardware, and where upstream support might be coming soon.

In regards to async, we’d ultimately have to complain at smart-leds upstream for this, as it needs a split API. I do actually have several other things I want to raise with them in regards to color-type generic drivers (see the upcoming 0.27 release, which makes the driver fully generic over color types, but with my own abstractions).

kleinesfilmroellchen commented:
> If I slipt the `async fn write()` into a creation part and a send part, we can create multiple futures with prepared data, and send them then all in sync.

Cool that this works with the trait API (didn’t know that you can just *not* use the async sugar if you want to, even if it’s declared that way by upstream). I’ll use that code for 0.27 after testing.

kleinesfilmroellchen commented:
My oscilloscope says your code is good, so it is going into 0.27!

TuxCoder commented:
Thanks for testing and merging!

TuxCoder commented:
> I think the memsize is a fine solution for these cases. At least on models like the C6 you can even increase it up to 4, which then also occupies the two receive channel buffers. But as we observed it’s a trade-off and not necessary on faster chips and fewer LEDs, which is why I made it an optional configuration. I assume that this worked for you, as released in 0.26?

Yes works perfectly.

> Cool that this works with the trait API (didn’t know that you can just not use the async sugar if you want to, even if it’s declared that way by upstream). I’ll use that code for 0.27 after testing.

Yes the async keyword is only synatx sugger around the `Future` trait 

> Fixed! GitHub appears to have done this automatically. Feel free to write up any issues you found on my repo.

Thx will do if I find something to improve / bugs.

Hardware limits are now discovered.
So I can start to program some animations :)
New chips coming soon, and will test with them / report bugs.
